File: 54-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        m, n = len(matrix), len(matrix[0])
        upper_bound, lower_bound = 0, m - 1
        left_bound, right_bound = 0, n - 1

        while (len(res) < m * n):
            # left -> right
            if (upper_bound <= lower_bound):
                for j in range(left_bound, right_bound+1):
                    res.append(matrix[upper_bound][j])
                upper_bound += 1
            # up -> down
            if (left_bound <= right_bound):
                for i in range(upper_bound, lower_bound+1):
                    res.append(matrix[i][right_bound])
                right_bound -= 1

            # right -> left
            if (upper_bound <= lower_bound):
                for j in range(right_bound, left_bound-1, -1):
                    res.append(matrix[lower_bound][j])
                lower_bound -= 1

            # down -> up
            if (left_bound <= right_bound):
                for i in range(lower_bound, upper_bound-1, -1):
                    res.append(matrix[i][left_bound])
                left_bound += 1
        return res

File: 54-spiral-matrix/solution.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int m = matrix.size(), n = matrix[0].size();
        int upper_bound = 0, lower_bound = m - 1;
        int left_bound = 0, right_bound = n - 1;

        while (res.size() < m * n) {
            // left -> right
            if (upper_bound <= lower_bound) {
                for (int j = left_bound; j <= right_bound; j++) {
                    res.push_back(matrix[upper_bound][j]);
                }
                upper_bound++;
            }
            // up -> down
            if (left_bound <= right_bound) {
                for (int i = upper_bound; i <= lower_bound; i++) {
                    res.push_back(matrix[i][right_bound]);
                }
                right_bound--;
            }

            // right -> left
            if (upper_bound <= lower_bound) {
                for (int j = right_bound; j >= left_bound; j--) {
                    res.push_back(matrix[lower_bound][j]);
                }
                lower_bound--;
            }

            // down -> up
            if (left_bound <= right_bound) {
                for (int i = lower_bound; i >= upper_bound; i--) {
                    res.push_back(matrix[i][left_bound]);
                }
                left_bound++;
            }
        }
        return res;
    }
};